### PR TITLE
bt-86: Sign-up

### LINKED ISSUE #86: Sign-up
Finilize please sign-up process

- [x] check and implement all TODO in sign-up process (password hash, ... etc)
- [x] check sign-up flow

### PR DESCRIPTION
**Pull request by Monyster**
Closes #86 

Check flow:

1. No body
```
POST /api/v1/auth/sign-up HTTP/1.1
Postman-Token: 5d81e8a4-14ab-45da-ab41-630f36205daf
Host: localhost:3001
Content-Length: 0
 
HTTP/1.1 422 Unprocessable Entity
content-type: application/json; charset=utf-8
content-length: 140
Date: Tue, 29 Aug 2023 15:36:45 GMT
Connection: keep-alive
Keep-Alive: timeout=72
 
{"errorType":"VALIDATION","message":"\"value\" must be of type object","details":[{"path":[],"message":"\"value\" must be of type object"}]}
```
2. Not all parameters in body
```
POST /api/v1/auth/sign-up HTTP/1.1
Content-Type: application/json
Postman-Token: 514c5015-446d-4e1e-822e-d7adf5a134f9
Host: localhost:3001
Content-Length: 36
 
{
"email": "qwerty@mail.com"
}
 
HTTP/1.1 422 Unprocessable Entity
content-type: application/json; charset=utf-8
content-length: 207
Date: Tue, 29 Aug 2023 15:39:34 GMT
Connection: keep-alive
Keep-Alive: timeout=72
 
{"errorType":"VALIDATION","message":"\"password\" is required. \"role\" is required","details":[{"path":["password"],"message":"\"password\" is required"},{"path":["role"],"message":"\"role\" is required"}]}
```
4. Invalid json format ( "," after password)
```
POST /api/v1/auth/sign-up HTTP/1.1
Content-Type: application/json
Postman-Token: bc96584c-dc0f-4ec0-b2c0-4a2784a6cb0d
Host: localhost:3001
Content-Length: 90
 
{
"email": "test@mail.com",
"role": "talent",
"password": "password",,,,
}
 
HTTP/1.1 500 Internal Server Error
content-type: application/json; charset=utf-8
content-length: 76
Date: Tue, 29 Aug 2023 15:40:51 GMT
Connection: keep-alive
Keep-Alive: timeout=72
 
{"errorType":"COMMON","message":"Unexpected token , in JSON at position 84"}
```
5. Validation email - wrong email
```
POST /api/v1/auth/sign-up HTTP/1.1
Content-Type: application/json
Postman-Token: b770cc22-694a-4f02-b66b-dee1eda24df5
Host: localhost:3001
Content-Length: 86
 
{
"email": "test.mail.com",
"role": "talent",
"password": "password"
}
 
HTTP/1.1 422 Unprocessable Entity
content-type: application/json; charset=utf-8
content-length: 111
Date: Tue, 29 Aug 2023 15:41:12 GMT
Connection: keep-alive
Keep-Alive: timeout=72
 
{"errorType":"VALIDATION","message":"Email is wrong","details":[{"path":["email"],"message":"Email is wrong"}]}
```
6. Validation role - bad role value
```
POST /api/v1/auth/sign-up HTTP/1.1
Content-Type: application/json
Postman-Token: f9683177-eb76-4dd2-9642-a6280e553ad2
Host: localhost:3001
Content-Length: 88
 
{
"email": "test@mail.com",
"role": "talenttt",
"password": "password"
}
 
HTTP/1.1 422 Unprocessable Entity
content-type: application/json; charset=utf-8
content-length: 180
Date: Tue, 29 Aug 2023 15:45:36 GMT
Connection: keep-alive
Keep-Alive: timeout=72
 
{"errorType":"VALIDATION","message":"\"role\" must be one of [talent, employer, admin]","details":[{"path":["role"],"message":"\"role\" must be one of [talent, employer, admin]"}]}
```
7. Validation password - not empty
```
POST /api/v1/auth/sign-up HTTP/1.1
Content-Type: application/json
Postman-Token: 0bd2ad5f-d64e-4a42-95d3-856284582bbe
Host: localhost:3001
Content-Length: 78
 
{
"email": "test@mail.com",
"role": "talent",
"password": ""
}
 
HTTP/1.1 422 Unprocessable Entity
content-type: application/json; charset=utf-8
content-length: 164
Date: Tue, 29 Aug 2023 15:46:00 GMT
Connection: keep-alive
Keep-Alive: timeout=72
 
{"errorType":"VALIDATION","message":"\"password\" is not allowed to be empty","details":[{"path":["password"],"message":"\"password\" is not allowed to be empty"}]}
```
8. Succesfull sign up
```
POST /api/v1/auth/sign-up HTTP/1.1
Content-Type: application/json
Postman-Token: 613300ab-da67-45c1-8cc8-c6c31213e048
Host: localhost:3001
Content-Length: 86
 
{
"email": "test@mail.com",
"role": "talent",
"password": "password"
}
 
HTTP/1.1 201 Created
content-type: application/json; charset=utf-8
content-length: 267
Date: Tue, 29 Aug 2023 15:46:25 GMT
Connection: keep-alive
Keep-Alive: timeout=72
 
{"id":"64ac6d4d-39c6-4d2a-adbc-0d39481c1046","email":"test@mail.com","role":"talent","token":"eyJhbGciOiJIUzI1NiJ9.eyJpZCI6IjY0YWM2ZDRkLTM5YzYtNGQyYS1hZGJjLTBkMzk0ODFjMTA0NiIsImlhdCI6MTY5MzMyMzk4NSwiZXhwIjoxNjkzNDEwMzg1fQ.aVj5vlew0yVfcJkmGwF5P62ht_hvT-IVAWoz0yZnZSU"}
```
9. User already exist
```
POST /api/v1/auth/sign-up HTTP/1.1
Content-Type: application/json
Postman-Token: 04861a0d-699f-405f-8816-ce343be28b17
Host: localhost:3001
Content-Length: 86
 
{
"email": "test@mail.com",
"role": "talent",
"password": "password"
}
 
HTTP/1.1 400 Bad Request
content-type: application/json; charset=utf-8
content-length: 58
Date: Tue, 29 Aug 2023 15:46:41 GMT
Connection: keep-alive
Keep-Alive: timeout=72
 
{"errorType":"COMMON","message":"Email is already taken."}
```

image

### CODE REVIEW

**Comment by nikita-remeslov** [comment]
@ptarho check please your part

**Review by MChornomaz — APPROVED** [review]

File: shared/src/bundles/users/types/user-sign-up-response-dto.type.ts [hunk 1]
```diff
@@ -1,6 +1,11 @@
+import { type ValueOf } from '~/types/value-of.type.js';
+
+import { type UserRole } from '../enums/enums.js';
+
 type UserSignUpResponseDto = {
     id: string;
     email: string;
+    role: ValueOf<typeof UserRole>;
```
Comment (ptarho) [inline shared/src/bundles/users/types/user-sign-up-response-dto.type.ts:8]: @nikita-remeslov

Comment (ptarho) [reply shared/src/bundles/users/types/user-sign-up-response-dto.type.ts:8]: Should we leave it for now?

**Review by ptarho — CHANGES_REQUESTED** [review]

File: shared/src/bundles/users/types/user-sign-up-response-dto.type.ts (hunk 1 above)
Comment (Monyster) [reply shared/src/bundles/users/types/user-sign-up-response-dto.type.ts:8]: I thought it need to choose a registration strategy for a new user.
But now I think that frontend already contains this.
So maybe it's useless value in response.

**Comment by Monyster** [comment]
@nikita-remeslov 
Fixed conflicts.

**Review by ptarho — APPROVED** [review]

**Review by anton-otroshchenko — APPROVED** [review]

**Comment by Monyster** [comment]
Fixed conflicts.

**Review by nikita-remeslov — APPROVED** [review]

**Pull request merged**